File: autonomy/camera-radar-fusion/src/shared/fusion/services/associate_camera_radar_detections.py
import logging

import numpy as np
from shared.camera.types.camera_detection import CameraDetection
from shared.fusion.utils.calculate_distance_to_bounding_box import (
    calculate_distance_to_bounding_box,
)
from shared.radar.types.radar_detection import RadarDetection

logger = logging.getLogger(__name__)


def associate_camera_radar_detections(
    camera_detections: list[CameraDetection],
    radar_detections: list[RadarDetection],
    distance_threshold_pixels: float = 100.0,
) -> list[tuple[int, int]]:
    """
    Associate camera and radar detections using spatial proximity.

    Uses greedy nearest-neighbor matching based on:
    1. Radar points inside camera bounding boxes (preferred)
    2. Minimum distance to bounding box edges

    Args:
        camera_detections: List of camera detections
        radar_detections: List of radar detections with image coordinates
        distance_threshold_pixels: Maximum distance for association

    Returns:
        List of matched pairs as (camera_index, radar_index) tuples

    """
    if not camera_detections or not radar_detections:
        return []

    # Build distance matrix
    distance_matrix = np.zeros((len(camera_detections), len(radar_detections)))

    for camera_index, camera_detection in enumerate(camera_detections):
        for radar_index, radar_detection in enumerate(radar_detections):
            distance = calculate_distance_to_bounding_box(
                radar_detection,
                camera_detection,
            )
            distance_matrix[camera_index, radar_index] = distance

    # Greedy matching: assign each radar to closest camera
    matched_pairs = []
    used_camera_indices = set()
    used_radar_indices = set()

    # Sort by distance (closest first)
    camera_indices, radar_indices = np.where(
        distance_matrix <= distance_threshold_pixels,
    )
    distances = distance_matrix[camera_indices, radar_indices]
    sorted_indices = np.argsort(distances)

    for index in sorted_indices:
        camera_index = camera_indices[index]
        radar_index = radar_indices[index]

        # Skip if already matched
        if camera_index in used_camera_indices or radar_index in used_radar_indices:
            continue

        matched_pairs.append((camera_index, radar_index))
        used_camera_indices.add(camera_index)
        used_radar_indices.add(radar_index)

    logger.debug(
        f"Associated {len(matched_pairs)} pairs from {len(camera_detections)} "
        f"camera and {len(radar_detections)} radar detections",
    )

    return matched_pairs

File: autonomy/camera-radar-fusion-rust/src/shared/fusion/services/associate_camera_radar_detections.rs
use crate::shared::camera::types::camera_detection::CameraDetection;
use crate::shared::fusion::utils::calculate_distance_to_bounding_box::calculate_distance_to_bounding_box;
use crate::shared::radar::types::radar_detection::RadarDetection;
use std::collections::HashSet;
use tracing::debug;

pub fn associate_camera_radar_detections(
    camera_detections: &[CameraDetection],
    radar_detections: &[RadarDetection],
    distance_threshold_pixels: f64,
) -> Vec<(usize, usize)> {
    if camera_detections.is_empty() || radar_detections.is_empty() {
        return Vec::new();
    }

    let mut distance_pairs: Vec<(f64, usize, usize)> = Vec::new();

    for (camera_index, camera_detection) in camera_detections.iter().enumerate() {
        for (radar_index, radar_detection) in radar_detections.iter().enumerate() {
            let distance = calculate_distance_to_bounding_box(radar_detection, camera_detection);

            if distance <= distance_threshold_pixels {
                distance_pairs.push((distance, camera_index, radar_index));
            }
        }
    }

    distance_pairs.sort_by(|a, b| a.0.partial_cmp(&b.0).expect("Failed to compare distances"));

    let mut matched_pairs: Vec<(usize, usize)> = Vec::new();
    let mut used_camera_indices: HashSet<usize> = HashSet::new();
    let mut used_radar_indices: HashSet<usize> = HashSet::new();

    for (_distance, camera_index, radar_index) in distance_pairs {
        if !used_camera_indices.contains(&camera_index)
            && !used_radar_indices.contains(&radar_index)
        {
            matched_pairs.push((camera_index, radar_index));
            used_camera_indices.insert(camera_index);
            used_radar_indices.insert(radar_index);
        }
    }

    debug!(
        "Associated {} pairs from {} camera and {} radar detections",
        matched_pairs.len(),
        camera_detections.len(),
        radar_detections.len()
    );

    matched_pairs
}
